### PR TITLE
Feature improve box queries and fix problems with rotations

### PR DESCRIPTION
**Pull request by petterarvidsson**
- Add box shape queries that makes it easier to ask for volumes of blocks of different shapes
- Add generic method to send events to plugins listening via universe
- Fix problems introduced code for translating rotations (figuring out which face was clicked on a rotated block)
- Add some generic helper methods and improve pretty printing

### CODE REVIEW

**Comment by petterarvidsson** [comment]
@nsg Lots of clean-ups and new features.

**Comment by petterarvidsson** [comment]
@nsg I think this stuff is ready now. Do you want to take a look at the code and more importantly the doc?

**Comment by nsg** [comment]
@petterarvidsson sure! Give me a hour or so.

File: src/main/java/konstructs/api/Box.java [hunk 1]
```diff
@@ -208,6 +177,14 @@ public int arrayIndex(Position p) {
         return arrayIndexLocal(p.subtract(from));
     }
 
+    public Position arrayIndexAsPosition(int index) {
```
Comment (nsg) [inline src/main/java/konstructs/api/Box.java]: A public function with no docs

File: src/main/java/konstructs/api/InclusiveBox.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package konstructs.api;
+
+/**
+ * Class that represents a {@link Box box} starting at a certain position and
+ * ending at another certain position.
+ * It differs from {@link Box box} (which it also extends) in the sense that
+ * end is inclusive and start maybe be smaller than end in all dimensions.
```
Comment (nsg) [inline src/main/java/konstructs/api/InclusiveBox.java]: I understand what this sentence says, but it's a little confusing :)

File: src/main/java/konstructs/api/Orientation.java [hunk 3]
```diff
@@ -58,24 +58,42 @@ public Rotation getRotation() {
...
      * @param dir The direction to be translated
      * @return The translated direction
      */
-    public Direction translateDirection(Direction dir) {
-        return Direction.get(getRotationMatrix().multiply(dir.getVector()));
+    public Direction translateFacePointedAt(Direction dir) {
+        return Direction.get(getFacePointedAtRotationMatrix().multiply(dir.getVector()));
+    }
+
+    public Direction translateFacePointingIn(Direction dir) {
```
Comment (nsg) [inline src/main/java/konstructs/api/Orientation.java]: No docs

File: src/main/java/konstructs/api/Orientation.java [hunk 4]
```diff
@@ -58,24 +58,42 @@ public Rotation getRotation() {
...
+    }
+
+    /**
+     * Translate (rotate) a face pointed at using this orientation.
+     * If the direction is seen as a face pointed at, then this method returns the direction of the face translated
+     * by this orientation. E.g. if passed the upward direction (block was pointed at from above) and the orientation
+     * has a direction of downward then face downward will be returned, since it is pointing up.
+     *
+     * This means that this method can be used when one want to know which face that was interacted with when a block
+     * is rotated.
```
Comment (nsg) [inline src/main/java/konstructs/api/Orientation.java:89]: I guess these are fine, but I'm not saying that it's easy to understand. I like this "example of usage" here, maybe have more of these in hard to understand methods :)

**Review by nsg — CHANGES_REQUESTED** [review]
Over all really nice, as soon we have this merged I will test it with water.

**Comment by petterarvidsson** [comment]
@nsg Pushed some documentation updates. Please have a look at them!

**Review by nsg — APPROVED** [review]
Looks good

**Pull request merged**